[soft] Meta-annotate using @ExtendWith

kriegfrj opened the pull request:
From Jupiter 5.8 and on, this allows Jupiter soft assertion tests to automatically register the `SoftAssertionsExtension` simply by use of the `@InjectSoftAssertions` field.

At least, that's the theory. The unit test I put in place to test it doesn't seem to be working. Hailing @bjhargrave (who added this feature to [osgi-test](https://github.com/osgi/osgi-test)) to see if he can shed any light on this issue.

kriegfrj committed:
[soft] Meta-annotate using @ExtendWith

From Jupiter 5.8 and on, this allows Jupiter soft assertion
tests to automatically register the SoftAssertionsExtension
simply by use of the @InjectSoftAssertions field.

bjhargrave commented:
I am not sure what is going on here. I built the assertj-core snapshot with the change to `InjectSoftAssertions` and tried to use it in the osgi-test build after removing the `@ExtendWith(SoftAssertionsExtension.class)` in `PromiseAssertTest`. It fails in the same way with the field injection not happening.

I then looked at how the two extensions, `SoftAssertionsExtension` and `BundleContextExtension`, do the field injection. In `BundleContextExtension`, we do the field injection during the `BeforeEachCallback` (and `BeforeAllCallback`) callback. In `SoftAssertionsExtension`, it appears to do the field injection during the `TestInstancePostProcessor` callback. So I am not sure if this is something about how JUnit's support of `@ExtendWith` on the annotation is working or not.

kriegfrj commented:
Thanks for the insight, @bjhargrave. I'll try digging further. Just wanted to make sure I hadn't missed something obvious.

kriegfrj commented:
> I then looked at how the two extensions, `SoftAssertionsExtension` and `BundleContextExtension`, do the field injection. In `BundleContextExtension`, we do the field injection during the `BeforeEachCallback` (and `BeforeAllCallback`) callback. In `SoftAssertionsExtension`, it appears to do the field injection during the `TestInstancePostProcessor` callback. So I am not sure if this is something about how JUnit's support of `@ExtendWith` on the annotation is working or not.

This is the key. `SoftAssertionsExtension` uses the `TestInstancePostProcessor` callback so that it can transparently handle `PER_CLASS` + `CONCURRENT` (something we have not concerned ourselves with in osgi-test) using `ThreadLocal` under-the-hood. The injection is done in two phases - the TIPP callback initialises & injects the `SoftAssertionsProvider` instances, and `beforeEach()` sets the delegate of each `SoftAssertionsProvider` to the common collector.

When you use implicit registration, it seems that the TIPP callback is not being called.

I'm trying to find a workaround but it's not easy...

bjhargrave commented:
> When you use implicit registration, it seems that the TIPP callback is not being called.

Perhaps this is a bug in JUnit? Or an undocumented restriction?

kriegfrj commented:
> > When you use implicit registration, it seems that the TIPP callback is not being called.
> 
> Perhaps this is a bug in JUnit? Or an undocumented restriction?

I think definitely it is one of those two...

kriegfrj commented:
> > > When you use implicit registration, it seems that the TIPP callback is not being called.
> > 
> > 
> > Perhaps this is a bug in JUnit? Or an undocumented restriction?
> 
> I think definitely it is one of those two...

Or perhaps a third option: it is a *documented* restriction:

From [section 5.2.1](https://junit.org/junit5/docs/current/user-guide/#extensions-registration-declarative):
> @ExtendWith fields may be either static or non-static. The documentation on Static Fields and Instance Fields for @RegisterExtension fields also applies to @ExtendWith fields.

The relevant documentation for `@RegisterExtension` [here](https://junit.org/junit5/docs/current/user-guide/#extensions-registration-programmatic-instance-fields): 
>If a @RegisterExtension field is non-static (i.e., an instance field), the extension will be registered after the test class has been instantiated and after each registered TestInstancePostProcessor has been given a chance to post-process the test instance (potentially injecting the instance of the extension to be used into the annotated field). Thus, if such an instance extension implements class-level or instance-level extension APIs such as BeforeAllCallback, AfterAllCallback, or TestInstancePostProcessor, those APIs will not be honored. 

So a workaround is required.

At this point, I'm extremely thankful we took the time to implement thorough tests - the rewrite required is substantial and the tests make sure that I haven't broken any of the (many) weird corner cases.  I think I have my workaround nearly ready.

kriegfrj committed:
[soft] Improved test robustness

Signed-off-by: Fr Jeremy Krieg <fr.jkrieg@greekwelfaresa.org.au>

kriegfrj committed:
[woft] Changed to make implicit registration work

Due to a limitation in Jupiter, extensions registered via @ExtendWith
annotated on a field won't have the postInstance callback called.
So we do the initialisation inside beforeEach() instead.

Signed-off-by: Fr Jeremy Krieg <fr.jkrieg@greekwelfaresa.org.au>

kriegfrj commented:
This latest push passes all the tests (including the new one I added to check that the meta-annotation is working). There might be some obscure problems with nested tests that mix PER_CLASS and PER_METHOD, but other than that it should be fine.

scordio commented:
Thanks a lot to both of you for your effort!

If you have time, would you mind expressing your opinion on #2356?
